### ISSUE TITLE
User config and data migration across branches

### ISSUE THREAD

**Issue by irydacea** [wesnoth/wesnoth#2710]
One common complaint about Wesnoth's development approach is that players are expected to lose all their user configuration and data (add-ons, saves) when upgrading from one stable branch to the next one, or migrate them by hand and brace for the issues that can result from:

* Trying to run Wesnoth with preferences from an older version (which *usually* works just fine, but issue #2612 is an example of how this can go wrong due to oversights)
* Using obsolete add-ons (which usually have been updated to the newer stable branch, but the player has to go to the Add-ons Manager to reinstall them by hand or upgrade them at once if they copied over their `data/add-ons` dir themselves)
* Loading older saves (start-of-scenario saves *usually* work just fine, but this is not always the case if the attached WML in units has changed in the meantime, e.g. abilities and other data that isn't defined/overridden by the engine)

Even though it's easy to shrug it off and just decide that the clean-slate approach for new stable branches is part of the project's development philosophy somehow, once Wesnoth is released on Steam we can expect more vocal members of the audience (e.g. game reviewers) to complain rather loudly once the 1.16 stable series lands and people suddenly have lost all their gameplay progress and profile configuration from 1.14 unless they go out of their way to read instructions on a wiki/forum post and do file management operations to get their stuff back -- or just give up, depending on the case.

So, in my opinion, going forward, some form of automated or semi-automated migration task will be necessary to avoid this issue from cropping up every time and demotivating players who've spent countless hours playing the game to get 100% SP campaign completion or what have you.

A potential implementation for this would be as follows:

* On first start-up, if Wesnoth detects its user config and data dir are **both** empty or non-existent (obvious sign of the first launch of the new branch), it should look for the dir(s) for the previous stable branch and offer the user the option to migrate their profile from the previous version. If the user declines, then the game continues with a clean slate like it currently does.
* If the user opts to migrate their profile, then saves, preferences, and credentials are copied (or moved?) over to the new version's profile. The game should then gather a list of add-ons installed on the old version (currently this is a trivial directory enumerate operation), connect to the add-ons server, and see which add-ons are available for the new version and offer automatically downloading them.
* Before finishing and loading the preferences file, the game should perform any syntax/structure changes that may be required on the new profile's copy of it. If for some reason the process fails, inform the user accordingly and just proceed with fresh preferences (but this should never happen).
* Saves are trickier because they can involve syntax changes that are defined by the engine (e.g. the structure of replays, how units are stored in the gamestate snapshot, etc.) and changes that are defined by Lua/WML. Some API for applying transforms to save files á la `wmllint` would need to be built into the engine, and those transforms would be applied when loading tainted saves according to the version of the game that generated them. The catch here is that this means that developers can no longer break WML/Lua API compatibility willy-nilly and a more formal transition process would be required for every API break.

This is an issue that at least concerns Windows (which uses a combined user config+data dir with a version identifier in its name for each branch) and Linux (which uses the version identifier for the user data dir by default but *not* for the cache and config dirs). I do not know anything about how Wesnoth manages user resources on macOS to suggest how the migration would work on that platform. Builds using a custom user config+data dir path (e.g. using SCons `prefsdir` option) can probably just have the migration code disabled by default since in that particular case the assumption ought to be that the user knows what they're doing -- unless there's some Linux distribution making use of said option for stock packages for some weird reason.

**Comment by shikadiqueen** [comment]
I see some potential for abuse with a system like that. For example, since I haven't uploaded my campaigns to 1.14 yet, someone could take the chance to upload their own add-ons posing as mine in the migration system's view, thereby causing users of my add-ons to download something unrelated.

Granted, that kind of thing can simply be explicitly outlawed in the server rules and then we can just ask players to report any incidents like that.

**Comment by GregoryLundberg** [comment]
Yes. Signing keys solve myriad issues, including handling the password (which could go away entirely with signing keys).

The entire idea has merits, it just needs to be decomposed into bits we can do sooner and those which need more work.

**Comment by shikadiqueen** [comment]
The signing keys stuff reeks of overengineering and I doubt the average add-on author will understand it right away.

(Also, somewhat off-topic for this issue, but my plan is to work on enabling campaignd to use forum authentication before 1.15.0 comes out.)

**Comment by CelticMinstrel** [comment]
I think the keys thing could theoretically be made transparent to the user if the game generates a key-pair when you first upload a new addon? Not quite sure though.

Using forum credentials instead of an addon passcode does seem like a nicer approach but would require a way to change ownership. (Preferably other than deleting and reuploading.)

**Comment by stevecotton** [comment]
Alternative idea for the save files: leave them in the old directory, and add a button to the load-game dialogue that changes to that directory. I've put a prototype for this in 5aa2d54fa84e251a5cacebb40a055361ec5c1d6f, in my stevecotton:prototype_load_from_1_14 branch.

This doesn't fix any compatibility problems, but is better than dropping all old saves. It's also likely to be used by developers who currently manually copy files from 1.14 to 1.15.

**Comment by stevecotton** [comment]
> Saves are trickier because they can involve syntax changes that are defined by the engine (e.g. the structure of replays, how units are stored in the gamestate snapshot, etc.) and changes that are defined by Lua/WML. Some API for applying transforms to save files á la `wmllint` would need to be built into the engine, and those transforms would be applied when loading tainted saves according to the version of the game that generated them.

The framework for this already exists in savegame.cpp's convert_old_saves().

**Comment by stevecotton** [comment]
Mainly to @shikadiqueen and @nemaara : although renumbering scenarios makes sense, it will break both (start of renumbered scenario) and (during previous scenario) saves. Should we aim for compatibility, maybe by adding transition scenarios?

**Comment by nemaara** [comment]
Are you asking this in general, or for specific cases?

**Comment by stevecotton** [comment]
I'm asking in general.

I can see campaign-specific answers in some cases, for example UtBS from 1.12 to 1.14 completely changed the Quenoth elves and the way their units work together, so it seems unreasonable to try to maintain compatibility. On the other hand, DiD's Endless Night is something that a player might want to play occasionally with their overpowered army, the units for that army are still in mainline, and so compatibility for that specific scenario seems worth preserving.

In both of those cases, the campaign has changed so much that even players who only want to play new-to-them campaigns might want to play it again.

**Comment by nemaara** [comment]
In the case of campaigns (or relevant sections) that haven't changed a lot, probably makes sense to try to number scenarios in a way that maintains compatibility (see how doofus-01 did it for SoF). If a campaign did change significantly, then like you said, it makes sense to play it again.

**Comment by gfgtdf** [comment]
well what we should definiteley do is copying over (or even sharing) the preferences, this includes the completion states of the campaign etc.

Saves file conversion is very hard to do, and even if we somehow manage to have  an automated tool for the c++ changes, we still might have braking changes in the wml code, and with umc code that might store whatever in wml variables i don't think this is really possible.

An alterntive approach to that we could maybe do in particular wrt the steam release is maybe to simply not remove the old wesnoth files (data and binary) and make the wesnoth 1.16 client automaticially open old saves with the 1.14 clients when it detects them.

**Comment by AI0867** [comment]
Updating start-of-scenario saves would be a lot easier than in-scenario saves, but there are a lot of possible issues, especially if the campaign has been through a lot of changes:

- Does the scenario still exist?
- Does the difficulty level still exist?
- Does the player have all current important characters that he should have on his recall list?
- If the campaign has been rebalanced, do the carryover gold and recalls still make sense?
- Do the unit types in the save still exist, and do they correspond to the units that the campaign currently uses?

There are undoubtedly many cases where importing an old save would be simple, but I suspect there are also cases where the best solution would be to throw up our hands and tell the player they're better off restarting the campaign.

**Comment by gfgtdf** [comment]
> I don't know why you're worrying about WML variables. Those are by definition arbitrary data that a save file conversion process would need to leave untouched

Wml variables often contain unit data to be unstored or wml event code to be inserted via [insert_tag], if they make use of a code that is removed in a later version wesnoth n the code will fail. Even worse: The variables attributes can contain lua or formula code that make use of removed functions.

**Comment by gfgtdf** [comment]
> I also don't think(?) that it's possible to have both the 1.14 and 1.16 clients installed on Steam under the same app at the same time. Probably the closest we could do is have 1.14 available under the BETAS tab, which would at least more easily allow people to switch back and forth between versions.


The idea was basically to include the whole wesnoth 1.14 in the wesnoth 1.16 steam package, maybe in a subdirectory

**Comment by stevecotton** [comment]
> I also don't think(?) that it's possible to have both the 1.14 and 1.16 clients installed on Steam under the same app at the same time. Probably the closest we could do is have 1.14 available under the BETAS tab, which would at least more easily allow people to switch back and forth between versions.

Would it be better to treat 1.14 and 1.16 as separate games on Steam, as if  1.16 was a sequel to 1.14? (I don't know much about Steam, would appreciate details from people who do.)

Many of the features in this discussion would still be valuable, but this would avoid the problems of automatic upgrades from 1.14 to 1.16.

**Comment by shikadiqueen** [comment]
> Would it be better to treat 1.14 and 1.16 as separate games on Steam, as if 1.16 was a sequel to 1.14? (I don't know much about Steam, would appreciate details from people who do.)

I don't think that Valve would allow us to do that, and from a player's perspective it's still terrible UX.

**Comment by CelticMinstrel** [comment]
> Updating start-of-scenario saves would be a lot easier than in-scenario saves, but there are a lot of possible issues

That's true, but the most we can do is a "best effort" under the assumption that the scenario / campaign hasn't changed much. Perhaps there could also be some campaign / scenario versioning system to let the game know that conversion is doomed to fail.

> Wml variables often contain unit data to be unstored or wml event code to be inserted via [insert_tag], if they make use of a code that is removed in a later version wesnoth n the code will fail. Even worse: The variables attributes can contain lua or formula code that make use of removed functions.

Sure. They're still arbitrary data. I suppose since stored units are pretty common though it _might_ be possible to automatically detect and update them… but either way, probably the correct approach is a mechanism to pass the compatibility code on to the scenario developer, who knows what those variables hold and how they might need to be updated.

> The idea was basically to include the whole wesnoth 1.14 in the wesnoth 1.16 steam package, maybe in a subdirectory

I think this is an extremely bad idea.

> I also don't think(?) that it's possible to have both the 1.14 and 1.16 clients installed on Steam under the same app at the same time. Probably the closest we could do is have 1.14 available under the BETAS tab, which would at least more easily allow people to switch back and forth between versions.

I don't know if Steam _officially_ allows having two versions of the same app installed at the same time, but someone could easily install one version, rename the install folder, and then install the other version. I think the betas feature is the correct way to handle this in any case.

**Comment by hexagonrecursion** [comment]
> I think the keys thing could theoretically be made transparent to the user if the game generates a key-pair when you first upload a new addon?

I doubt it is possible to make cryptographic signatures *completely transparent* to the signer. Consider the following three scenarios:

# Deleting the user data folder
1. Alice uploads "My Example Add-on 1.0"
2. Bob installs "My Example Add-on 1.0"
2. Alice uploads "My Example Add-on 1.1" and everything magically works
3. Bob automatically upgrades to "My Example Add-on 1.1"
4. Alice experiences a game-breaking bug and attempts to fix it by deleting her user data folder and reinstalling the game
5. Now there is no way for Alice to prove that she is the same Alice as before

# Hard drive crash / OS corruption / ...
1. Carol uploads "Carol's Add-on 1.0"
2. Carol does not back his private key up because he does not even know that he has one
3. Carol experiences a hard disk failure

# Multiple devices
1. Dave has a laptop and a desktop
2. Dave uploads "Dave's Add-on 1.0" from his desktop
3. Dave tries to upload "Dave's Add-on 1.1" from his laptop and is surprised when that does not work

If you introduce a way for an add-on developer to get a new key by authenticating to a server (eg via login+password), you essentially create a roundabout way to authenticate uploads via login+password.

I don't think it is possible to make public key crypto *transparent*. Making it *user-friendly* could be achievable, but it is a hard problem. I would not recommend Wesnoth to attempt to tackle it on its own.

**Comment by stevecotton** [comment]
Pinging everyone who's subscribed to this - I've opened PR #5574 which will add the paragraph below to the confirm-loading-an-old-save dialog, based on discussion in #5517.

"<b>Warning:</b> for saves in the middle of campaigns, you should either use the old version or restart the campaign. Even when a saved game seems to load successfully, there are many ways in which it could go subtly wrong. The difficulty, the challenge, the <i>fun</i> may be missing."

**Comment by Pentarctagon** [comment]
For replays, while it would obviously be a lot of work, I wonder if it would be more practical to instead have replays be more imperative than interpretive like they are now.  And by that I mean, instead of the game engine needing to redo things like [making calls to the PRNG](https://forums.wesnoth.org/viewtopic.php?f=10&t=53824), just have there be commands that say directly what to do.  Essentially there would be no need to understand why or how something happened, the replay would simply re-execute the end result of each action.  There would then be little chance of replays breaking between versions, since there should be no way in which basic, simple commands like "unit at 5,5 moved to 6,7" change meaning.

**Comment by Pentarctagon** [comment]
In any case though, as far as copying the 1.14 preferences file and attempting to redownload add-ons, is anyone already planning to implement those or should I take a look at them?

**Comment by shikadiqueen** [comment]
I may do that within the next couple of weeks, actually.

**Comment by Pentarctagon** [comment]
Since Iris said on Discord that she won't be able to take a look into this, I plan to implement the following hopefully in the relatively near future:
* Prompting whether to migrate their profile.  This will require a string change after the string freeze has taken effect, but it would be just a single sentence, and I think this is important enough to make an exception.
* Copying the preferences file.  I don't know what, if any, syntax changes there have been since 1.14, and I don't know how I would check for that either however.
* Copying the credentials file (and maybe make it work for multiple usernames).
* Automatically redownloading add-ons from the 1.16 add-ons server, if they exist.

**Comment by stevecotton** [comment]
Something that's coming up on IRC/Discord today - there's two migrations to worry about:
* 1.14 -> 1.16
* Steam 1.15 beta users -> 1.16

**Comment by Pentarctagon** [comment]
Both are identical aside from the source directory to look in for preferences etc, I would think?

**Comment by Pentarctagon** [comment]
Removing from the 1.16.0 milestone since what can realistically be addressed for 1.16.0 has been addressed.

**Comment by cooljeanius** [comment]
> Removing from the 1.16.0 milestone since what can realistically be addressed for 1.16.0 has been addressed.

So... the "Blocker: New Stable" label can go, too, then, can't it? Or do you leave that on?

**Comment by Pentarctagon** [comment]
I'd like to have some sort of answer for everything here by 1.18, even if for some things that can break save/replay compatibility the answer is we just don't have a good solution.

**Comment by cooljeanius** [comment]
> I'd like to have some sort of answer for everything here by 1.18, even if for some things that can break save/replay compatibility the answer is we just don't have a good solution.

oh right, that reminds me, I'd like to see if we can get the patch from this forum thread updated and merged: https://forums.wesnoth.org/viewtopic.php?f=10&t=53824&start=15

**Comment by Pentarctagon** [comment]
Someone would of course need to first submit the patch for inclusion.

**Comment by cooljeanius** [comment]
> Someone would of course need to first submit the patch for inclusion.

I've created a branch for working on the patch, but unfortunately trying to apply it results in a rather large rejects file that I'm not sure what to do with: https://github.com/cooljeanius/wesnoth/blob/replay_old_saves/src/savegame.cpp.rej

**Comment by Pentarctagon** [comment]
Unfortunately I've only ever done things via PR rather than patch files, so I can't really help with that.

**Comment by Wedge009** [comment]
I like what has been implemented to date with regards to having access to save files across other versions. What are the subtle things left to do for this report (the thread is rather long to read through everything).

> "Warning: for saves in the middle of campaigns, you should either use the old version or restart the campaign..."

Thinking about Steam users - will they get a choice to delay updating to 1.16.0 when that gets released or will it be forced on them? I understand 1.15.x is opt-in via the beta channel, but the major update to 1.16.0 would happen automatically (if the user has automatic updates enabled, which they are by default).

**Comment by Pentarctagon** [comment]
The remaining piece is some sort of built-in way to handle old syntax in saves/replays.

They would be updated to 1.16.0 automatically, but 1.14 would be available still under the BETAS setting.

**Comment by stevecotton** [comment]
Would it make sense to upload 1.16.0 to Steam's beta channel so that players get until 1.16.1 to schedule their upgrade, along with a post on the Steam forums saying that:

* 1.16.0 is the release,
* it's recommended to finish your current campaign and then update before starting a new one,
* the only planned differences between 1.16.0 and 1.16.1 are some translation tweaks

We know that (hopefully) 1.15.18 going to be similar to 1.16.0, but I think people are more likely to upgrade when the big number changes. Also, 1.15.18 will still have a different downloads folder to 1.16.

**Comment by irydacea** [comment]
Why would we need to have people "schedule" their upgrades all of a sudden? I've not heard of any other game doing any such thing, and you might be overestimating people's ability to find out this information through sources other than the game UI itself.

**Comment by Pentarctagon** [comment]
@stevecotton I'm not really seeing a reason to wait to do most of those things until 1.16.0?  Announcing 1.16.0's imminent release and how to handle going back to 1.14 if needed is something that should be done prior rather than after.

**Comment by kabachuha** [comment]
> along with a post on the Steam forums saying that:

>  1.16.0 is the release,
>  it's recommended to finish your current campaign and then update before starting a new one,
>  the only planned differences between 1.16.0 and 1.16.1 are some translation tweaks

@stevecotton I think we are going to have a very similar message in the Release Candidate (which will also still stay in the betas) announcement except for '1.16.0' there standing our RC1 and for '1.16.1' the release version itself, so the exact version numbering won't change the already existing meaning.  The players shouldn't be ininformed if they read the update notes.

**Comment by stevecotton** [comment]
Because of https://github.com/wesnoth/wesnoth/issues/5517#issuecomment-774415416 , combined with our campaigns being long enough for occasional players to spend more than a week of calendar time on them. Of other games, I can offer the anecdote of RimWorld streamers stating that they're using a particular version for a given "let's play" series.

While there's a lot of people that won't find out until after the automatic upgrade, the first post in this issue is about "once Wesnoth is released on Steam we can expect more vocal members of the audience (e.g. game reviewers) to complain rather loudly once the 1.16 stable series lands". Those very same vocal members can get the word out :D

@kabachuha this isn't about a 15.x to 16.0 upgrade, this is about players upgrading from 1.14.17.

**Comment by kabachuha** [comment]
@stevecotton Maybe release a tiny final 1.14.18 update with nothing new except from a notice of the immenient 1.16 release in N days?

**Comment by irydacea** [comment]
> While there's a lot of people that won't find out until after the automatic upgrade, the first post in this issue is about "once Wesnoth is released on Steam we can expect more vocal members of the audience (e.g. game reviewers) to complain rather loudly once the 1.16 stable series lands". Those very same vocal members can get the word out :D

There's no difference between doing it for a release candidate or for 1.16.0 except that the latter is going to get us more **"where is 1.16.0 i removed and reinstalled the game on steam but i still see 1.14.17" posts everywhere.

**Comment by cooljeanius** [comment]
I have like a dozen or so campaigns going simultaneously on 1.14; I dunno if I'll be able to finish them all in time... I guess I'll just have to accept that I'll have to leave some of them unfinished...

**Comment by Pentarctagon** [comment]
There will be a 1.14 "beta" available for you to choose.  So you'll be automatically updated to 1.16, but you can immediately switch back to 1.14 if there are things you still want to do.

**Comment by Wedge009** [comment]
So it sounds like the main Steam release will be whatever we push it to be (currently 1.14.17, soon to be 1.16.0) and everything else will be opt-in via beta channel, both old and new. The 'old' part of that was something I didn't know about, but that makes sense, at least within the framework that Steam offers.

**Comment by Pentarctagon** [comment]
Also, to make sure it's said way ahead of time, my intention right now with the oldstable "beta" branches in steam is to keep them around for as long as their multiplayer and add-on server instances are available.  So 1.14 (beta) and 1.16 (default) will be available, then 1.16 (beta) and 1.18 (default), and so on.  I don't think it makes sense to try and use Steam's betas feature as a way to keep even older versions around indefinitely.

**Comment by Wedge009** [comment]
Fair enough - I assume any removal of old stable (so far only 1.14) versions would correspond with the shut-down of the MP and add-on servers. I also imagine that would be preceded by a Steam news update, with perhaps instructions for the traditional stand-alone installations for die-hards that really want to keep going on the old versions.

**Comment by Pentarctagon** [comment]
Yeah, it'd basically just be one more place to put the same information that's going to need to be posted for 1.12 a bit after 1.16.0 is released.

**Comment by AI0867** [comment]
> Also, to make sure it's said way ahead of time, my intention right now with the oldstable "beta" branches in steam is to keep them around for as long as their multiplayer and add-on server instances are available. So 1.14 (beta) and 1.16 (default) will be available, then 1.16 (beta) and 1.18 (default), and so on. I don't think it makes sense to try and use Steam's betas feature as a way to keep even older versions around indefinitely.

Games like Europa Universalis 4 do keep old versions around essentially indefinitely, for exactly the same reason: campaigns take a long time. Of course, they don't have separate add-on & mp servers that will get shut down, but if the long-lived games are mainly SP campaigns, that shouldn't really be a problem. A long-lived MP campaign could still be managed by manually distributing the add-ons when needed, and using the trunk MP server.

**Comment by Pentarctagon** [comment]
Having a beta option that identical to the default wouldn't make a lot of sense - it would just be a duplicate of the default.

**Comment by AI0867** [comment]
> Having a beta option that identical to the default wouldn't make a lot of sense - it would just be a duplicate of the default.

It would ensure that you don't accidentally update when a new stable is released.

**Comment by Pentarctagon** [comment]
It would make more sense to create that once the next stable branch is going to be released soon, rather than duplicating it for the entire stable series.

I also still disagree about keeping older stable versions available - we shouldn't have versions that are effectively unsupported and lack important functionality as such an easily available option.  If someone really wants to play such old versions of Wesnoth, then they can download a standalone version.

**Comment by AI0867** [comment]
Well, I'm not saying we should add them now, but if someone is still playing a campaign on the stable_1.14 beta channel several years from now, do we really want to forcefully upgrade him to 1.18 and make him download the non-steam version if he wants to continue his campaign?

**Comment by Pentarctagon** [comment]
Frankly, yes.  If someone wants to continue using unsupported versions of software, then they should have to go through a couple extra steps to do so, especially since we already know ahead of time that major features will be broken and the answer anyone will get regarding bugs will be "that version is unsupported, please use a newer version".

If there's a way to hide a beta branch so that only existing people using it will be able to continue using it, then that'd be fine too, but after 1.18 is released I don't want to have a bunch of Steam users complaining how 1.14 doesn't work anymore and we need to turn the multiplayer/add-ons servers back on.

**Comment by CelticMinstrel** [comment]
I think having the current, previous, and upcoming versions should be sufficient.

**Comment by Wedge009** [comment]
I'm not sure where we settled on announcing the imminent release of 1.16.0 and it replacing 1.14.17. I know at time of writing there's still at least a month to go, I am just asking the question because looking at the beta announcement in Steam there's nothing to indicate that 1.15.18 is considered a release candidate. Are we planning to warn/notify users, say, a week before planned final release that this is what's happening? I'm just thinking the average user playing on 1.14.17 may not pay all that much attention to the beta announcements if they've already chosen not to try the beta channel.

Just trying to pre-empt the questions like 'how can I continue my 1.14 campaign?' that would happen on/after release day otherwise.

**Comment by Pentarctagon** [comment]
There will be an announcement on the forums, discord, etc.

**Comment by Wedge009** [comment]
Would that include Steam users too? Just thinking that there are likely players who only use Steam and don't pay attention to the more traditional outlets.

I suppose we'll see what happens when the time comes, then.

**Comment by Pentarctagon** [comment]
The full list being: forums, Discord, Steam, itch.io, and maybe Twitter if it's something that can reasonably be fit into the character limit there.

**Comment by cooljeanius** [comment]
I'm doing the migration now myself... there's some text cut off here in this prompt that's part of the migration process:
<img width="646" alt="Screen Shot 2021-12-01 at 00 31 41" src="https://user-images.githubusercontent.com/1388660/144177564-034f0ebe-8853-418e-98c3-344c4978d800.png">
(i.e. the "Do you wish to install it before" seems incomplete)

**Comment by Pentarctagon** [comment]
Sounds like #3443

**Comment by cooljeanius** [comment]
Apostrophe not being rendered properly:
<img width="741" alt="Screen Shot 2021-12-01 at 01 49 49" src="https://user-images.githubusercontent.com/1388660/144185955-583b139a-01f8-4f89-b6ad-3309bf7e07d6.png">
(also note how multiple "install dependencies" dialogs are being stacked atop one another, instead of having older ones be cleared)

**Comment by Wedge009** [comment]
I'm thinking this isn't strictly related to the user config/data migration issue here.

Is it because the add-ons are titled with HTML-escaping?

The dialogue 'stacking' looks like it's a matter of the window not being refreshed as new dialogues are displayed. That particular issue has been reported before, but I can't find which one right now - or even if it's still open.

**Comment by cooljeanius** [comment]
So, I had a number of add-ons that Wesnoth reported as unavailable after the migration process:
<img width="842" alt="Screen Shot 2021-12-01 at 01 52 40" src="https://user-images.githubusercontent.com/1388660/144561779-0762abf6-dc69-4806-a1d2-07ea8c7388d4.png">
I'm wondering if it would be possible for Wesnoth to save this list, and then notify me when one of the add-ons that it was unavailable to migrate *does* become available?

**Comment by Pentarctagon** [comment]
Perhaps it could be stored in the preferences file, and then checked whenever the user updates add-ons.

**Comment by cooljeanius** [comment]
> Perhaps it could be stored in the preferences file, and then checked whenever the user updates add-ons.

should I open a separate issue about this, or can it stay part of this one?

**Comment by Pentarctagon** [comment]
Probably better as its own issue.

**Comment by cooljeanius** [comment]
> Probably better as its own issue.

ok, I opened #7070

**Comment by cooljeanius** [comment]
Another thing that I'd find helpful would be an easy way to re-run the migrator... should that get a separate issue as well?

**Comment by Pentarctagon** [comment]
That'd make sense.

**Comment by cooljeanius** [comment]
> That'd make sense.

OK, I opened #7936 for that.

**Comment by Pentarctagon** [comment]
Removed the `Blocker: New Stable` tag since, well, a stable release was done without this being fully resolved. I'm leaving it open since it would certainly be a nice thing to have, however I'm not sure the remaining parts could ever realistically be implemented given Wesnoth's current development approach.

**Comment by Wedge009** [comment]
Is there a quick summary of what's still outstanding for this issue? While specific to Steam, I mentioned this issue in #4163 as it seems to be a super-set of that specific issue.

**Comment by Pentarctagon** [comment]
I believe it's just the last point (saves compatibility).

**Comment by cooljeanius** [comment]
> I believe it's just the last point (saves compatibility).

That's something that @demario-w sometimes works on, isn't it?
